compiler/optimizer/optimizer.go

```diff
@@ -322,3 +340,144 @@ func pushDown(trunk *dag.Trunk) {
 	}
 	trunk.Pushdown = filter
 }
+
+// newRangePruner returns a new predicate based on the input predicate pred
+// that when applied to an input value (i.e., "this") with fields from/to, returns
+// true if comparisons in pred against literal values can for certain rule out
+// that pred would be true for any value in the from/to range.  From/to are presumed
+// to be ordered according to the order o.  This is used to prune metadata objects
+// from a scan when we know the pool-key range of the object could not satisfy
+// the filter predicate of any of the values in the object.
+func newRangePruner(pred dag.Expr, fld field.Path, o order.Which) *dag.BinaryExpr {
+	lower := &dag.This{Kind: "This", Path: field.New("from")}
+	upper := &dag.This{Kind: "This", Path: field.New("to")}
+	if o == order.Desc {
+		lower, upper = upper, lower
+	}
+	e, _ := buildRangePruner(pred, fld, lower, upper, o)
+	return e
+}
+
+// buildRangePruner creates a DAG comparison expression that can evalaute whether
+// a Zed value adhering to the From/To pattern can be excluded from a scan because
+// the expression "pred" would evalate to false for all values of "fld" in the
+// From/To value range.  If a pruning decision can be reliably determined then
+// the boolean return value is true; otherwise, false.
+func buildRangePruner(pred dag.Expr, fld field.Path, lower, upper *dag.This, o order.Which) (*dag.BinaryExpr, bool) {
```

Comment:
I don't see any way for this this function to return (nil, true), so I think you can simplify it by dropping the bool return parameter.